### PR TITLE
Demo data infrastructure

### PR DESCRIPTION
**Pull request by GenevieveBuckley**
# Description
This PR introduces a framework for distributing demo data to users. 
This PR adds three datasets (I wanted to stick to things that could download in <2 minutes, for now):
1. `cells3d` - from the scikit-image demo datasets
2. `mri` - MRI image of @haesleinhuepf's head, from [zenodo here](https://zenodo.org/record/4276076#.YYNnV5txWtm)
3. `plantseed` - fluorescence image of plant cell membranes, from [zenodo here](https://zenodo.org/record/4276076#.YYNnV5txWtm). 

Related to https://github.com/napari/napari/issues/316

Other datasets we would like to include in the future include:
- a tracking dataset (images and tracks, ideally)
- a geodata image
- a surface/mesh dataset
- lattice light sheet microscope data (either Talley's dataset, or a possibly synthetic smaller sized dataset). This one gets used heavily in tutorial examples.

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# References

- Pooch documentation: https://www.fatiando.org/pooch/latest/index.html
- scikit-image demo dataset subpackage: https://github.com/scikit-image/scikit-image/blob/c6535b3e1af6f44b18bcc2171a63c7d10243de07/skimage/data/_fetchers.py
 
# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

I've checked that I can download and open the images in ipython. I haven't added tests for this yet, because I'd like to discuss what the best use of our CI is (obviously it's not very useful to just spend CI time downloading lots of data). I guess we could test downloading one of the smallest datasets, like `cells3d`?

## Final checklist:
- [x] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

File: napari/data/__init__.py [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+    https://zenodo.org/record/5090508#.YYNYi5txWtm
+    """
+    fname = POOCH.fetch(
+        "Haase_MRT_tfl3d1.tif",
+        downloader=pooch.HTTPDownloader(progressbar=True),
+    )
+    return tifffile.imread(fname)
+
+
+def plantseed():
```
Comment (GenevieveBuckley) [inline napari/data/__init__.py:98]: Technically an ovule is an immature seed, so I'm not totally sure if it's accurate to call this "plantseed" or not. Any thoughts @haesleinhuepf?

**Comment by GenevieveBuckley** [comment]
Another point for discussion (and maybe a followup PR) is whether we should allow people to pass in an argument for which folder they want the dataset downloaded into. 

Right now it sticks data into a hidden folder `.cache/napari`. But if we're going to add really big datasets to the demo data then people are likely to not have enough space on their primary hard drive & might want to direct the download to an external hard drive.

I think we do want to make this possible and easy to do. I'd like to hear what other people think.

File: napari/data/__init__.py (hunk 1 above)
Comment (haesleinhuepf) [reply napari/data/__init__.py:98]: Huh, we should ask a plant scientist. Maybe @dlegland can help out?

**Comment by haesleinhuepf** [comment]
> Could we have a page in the napari documentation that lists the example datasets, their origin (and license) and some description? I was always curious for example what "cells3d" shows and who acquired it with what microscope and what markers...

I'm hoping that we make the docstrings detailed enough so that it will automatically build us that webpage. Is your suggestion that we add more detail to the docstrings (if yes, a template for what info should be included would be great to have suggested), or are you suggesting something else?

File: napari/data/__init__.py (hunk 1 above)
Comment (dlegland) [reply napari/data/__init__.py:98]: Hum, just saw my name in the discussion, but I'm not a plant scientist neither, while I work a lot with plant scientists! Concerning the question, I would say that the ovule is contained within the seed. I know that at least for wheat, embryo is one part amoung others of the grain, so this should hold for other seeds as well.

**Comment by haesleinhuepf** [comment]
Hey @GenevieveBuckley ,

don't edit my comments ;-)

> make the docstrings detailed enough so that it will automatically build us that webpage

Yes, great idea! I haven't checked the docstrings in detail. What should be in the docstrings may depend a lot on what data/source it is. I will make suggestions for updating the docstrings. :-)

File: napari/data/__init__.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    Notes
+    -----
+    This MRI dataset of a human head was provided by Robert Haase,
+    and acquired at University Hospital Carl Gustav Carus
+    of the University of Technology, TU Dresden
+    as part of academic training of students in the Department of Radiology.
+
+    This dataset, along with several others, has been made available on zenodo:
+    https://zenodo.org/record/5090508#.YYNYi5txWtm
+    """
```
Comment (haesleinhuepf) [inline napari/data/__init__.py:90]:
```suggestion

    The dataset is licensed by Robert Haase under a [CC-BY 4.0 license](https://creativecommons.org/licenses/by/4.0/legalcode)
    """
```
Just adding a license

**Comment by haesleinhuepf** [comment]
And one more idea: The MRI dataset could be resaved as uint16 and would then be half as large.

**Comment by tlambert03** [comment]
Quick question: any reason this needs to be added to core napari?  Seems like it could very easily be a napari plugin using the [sample data hookspec](https://napari.org/plugins/stable/hook_specifications.html#napari.plugins.hook_specifications.napari_provide_sample_data).  (And if we _really_ wanted to force everyone to have it, would could add that plugin to our dependencies)

**Comment by GenevieveBuckley** [comment]
> don't edit my comments ;-)

I'm sorry! ... I thought I clicked "Quote reply" from the dropdown :rofl:  That's embarrasing

**Comment by DragaDoncila** [comment]
> Quick question: any reason this needs to be added to core napari? Seems like it could very easily be a napari plugin using the [sample data hookspec](https://napari.org/plugins/stable/hook_specifications.html#napari.plugins.hook_specifications.napari_provide_sample_data). (And if we _really_ wanted to force everyone to have it, would could add that plugin to our dependencies)

I tend to agree this should be a plugin as it's exactly what the sample data hookspec was intended for. I also think we should add it as a dependency because it'd be great if we could provide more than just the `scikit-image` samples.

**Comment by haesleinhuepf** [comment]
> I also think we should add it as a dependency because it'd be great if we could provide more than just the `scikit-image` samples.

Where do these examples live btw? I mean, where is the hookspec in napari (?) or in which napari-plugin (?) implemented? I  recently searched for this and couldn't find it.

**Comment by tlambert03** [comment]
> Where do these examples live btw? I mean, where is the hookspec in napari (?) or in which napari-plugin (?) implemented? I recently searched for this and couldn't find it.

the examples are provided by scikit-image (via pooch), and it's a builtin hook implementation: https://github.com/napari/napari/blob/main/napari/plugins/_skimage_data.py

**Comment by tlambert03** [comment]
(you could add them there too if you want, I just don't think we should create an additional infrastructure when one already exists)

**Comment by sofroniewn** [comment]
> > Quick question: any reason this needs to be added to core napari? Seems like it could very easily be a napari plugin using the sample data hookspec. (And if we really wanted to force everyone to have it, would could add that plugin to our dependencies)
>
> I tend to agree this should be a plugin as it's exactly what the sample data hookspec was intended for. I also think we should add it as a dependency because it'd be great if we could provide more than just the scikit-image samples.

I think I agree with this too. We could make a nice `napari-sample-data` plugin that provided this stuff and then depend on that plugin if we want it to be shipped by default to everyone.

I think it will make development and review much easier and faster as it will be a much smaller and simpler repo.

What are your thoughts @GenevieveBuckley? You could start a plugin on your personal repo and then we could move it over to the main napari org, or we can create one right now on the main org?

Thoughts from @jni too?

Otherwise this looks great, and I'm excited to see it moving forward!!

**Comment by alisterburt** [comment]
I agree with the above, a nice `napari-sample-data` plugin (shipped with napari by default) would be great! It should definitely live on the org too (but could easily start on someones personal account)

**Comment by GenevieveBuckley** [comment]
Ok, makes sense. I'll close this and set up a `napari-sample-data` plugin instead.

**Comment by GenevieveBuckley** [comment]
A couple of other questions came up:
- Will the progress bar be visible to people who have selected a sample dataset from the GUI menu? Can you easily cancel it from the GUI if it is going to take too long?
- Giving people the option to download to a specific location: Is there a way to do this from the GUI?
- Robert suggested above that it was very important to have documentation pages describing the datasets. Is having that just as the docs for the plugin ok, or is that not discoverable enough?

**Comment by jni** [comment]
> Will the progress bar be visible to people who have selected a sample dataset from the GUI menu? Can you easily cancel it from the GUI if it is going to take too long?

I think this is a napari-plugin-engine-2 question and the answer is "eventually"?

> Giving people the option to download to a specific location: Is there a way to do this from the GUI?

This is plugin-specific, as I understand it. napari won't know where the data comes from, sample data plugins give arrays/array-likes. If those happen to be backed by a background download + local cache, napari does not know about this, and it becomes a plugin settings issue. e.g. right now the skimage data plugin has skimage call pooch which downloads data to the .cache/skimage folder.

> Robert suggested above that it was very important to have documentation pages describing the datasets. Is having that just as the docs for the plugin ok, or is that not discoverable enough?

I think the plugin docs is ok, in combination with work on napari-hub that will allow us to list plugins that provide sample data. Having said that I just saw that the scikit-image [API reference](https://scikit-image.org/docs/dev/api/api.html) *doesn't* provide documentation for the data module. 🤦 Raised an issue at scikit-image/scikit-image#6015

**Comment by GenevieveBuckley** [comment]
> This is plugin-specific, as I understand it. napari won't know where the data comes from, sample data plugins give arrays/array-likes. If those happen to be backed by a background download + local cache, napari does not know about this, and it becomes a plugin settings issue. e.g. right now the skimage data plugin has skimage call pooch which downloads data to the .cache/skimage folder.

Right, but if I configure my plugin to use pooch with a progress bar, is that actually visible anywhere? Even the "small" datasets take a minute or two, so it's necessary.

**Comment by GenevieveBuckley** [comment]
> Quick question: any reason this needs to be added to core napari? Seems like it could very easily be a napari plugin using the sample data hookspec. (And if we really wanted to force everyone to have it, would could add that plugin to our dependencies)

I notice we already have the scikit-image data bundled into core napari, so I'm wondering why this was considered so different.  https://github.com/napari/napari/blob/main/napari/plugins/_skimage_data.py

My original understanding of why the lightweight `napari_plugin_engine` package existed was so that other libraries (like skimage) could add this type of code to their own library (without having to depend on napari). Up until today, I thought that was what was happening here, but it's not.

**Comment by tlambert03** [comment]
> I notice we already have the scikit-image data bundled into core napari, so I'm wondering why this was considered so different.

Because this PR was creating a new infrastructure for accessing demo data.  Like I said in https://github.com/napari/napari/pull/3580#issuecomment-962107994, if you want to add these to our internal plugin, that's fine. 

We just shouldn't be introducing a new infrastructure for accessing demo data

**Comment by sofroniewn** [comment]
> > I notice we already have the scikit-image data bundled into core napari, so I'm wondering why this was considered so different.
>
> Because this PR was creating a new infrastructure for accessing demo data. Like I said in #3580 (comment), if you want to add these to our internal plugin, that's fine.

Yeah, we could add this code here https://github.com/napari/napari/tree/main/napari/plugins to something like `_pooch_data.py` or `_builtins_data.py` and have it use `napari_provide_sample_data` the same way the skimage one does, but i still think i'd rather us have `napari-sample-data` as it's own repo as it will be much easier to maintain in the long run.

> My original understanding of why the lightweight napari_plugin_engine package existed was so that other libraries (like skimage) could add this type of code to their own library (without having to depend on napari). Up until today, I thought that was what was happening here, but it's not.

I'd love it if we didn't have the skimage builtin plugin code here either (right now it's inside a try/ catch), but it seemed like a reasonable option until we get to a better future state. I'd probably want to move the skimage code over to the new `napari-sample-data`  repo too and eventually have no "builtin" plugin code in the main napari repo

**Pull request closed without merging**